### PR TITLE
[stdlib] Update Codable.swift (NFC)

### PR DESCRIPTION
**Pull request by benrimmington**
* Use `any` keyword.
* Use primary associated types.

### CODE REVIEW

**Comment by Azoy** [comment]
cc: @natecook1000

**Comment by natecook1000** [comment]
Some of these are kind of surprising (but my trust in the API digester is... not high)

> `+Func KeyedEncodingContainerProtocol.encode(_:forKey:) has generic signature change from <Self, T where Self : Swift.KeyedEncodingContainerProtocol, T : Swift.Encodable> to <Self where Self : Swift.KeyedEncodingContainerProtocol>`

This seems like it should be a totally standard change to be able to make safely — is it because `encode(_:forKey:)` is mutating that it drops `T` from the generic signature? How does it still work without that generic parameter? cc @hborla

> `return type change from Swift.Error? to (Swift.Error)?`

🧐

The changes that succeeded look good to me! 👍🏻

**Comment by glessard** [comment]
It does look like rdar://99932758, which is blocking https://github.com/apple/swift/pull/61088. There has been no additional information since #61088.

These changes are clearly source-compatible. The strange behaviour of the API digester suggests not trusting it, and manually checking the generated library before and after this for changes in the list of mangled symbols.

**Comment by glessard** [comment]
If we're sure it doesn't change the symbol list, then I would approve -- I haven't checked, and I would like to know whether anyone has. (I don't yet trust the automated tests surrounding these changes.)

**Comment by glessard** [comment]
@benrimmington I've "manually" checked the list of symbols, and I'm happy with what I see (i.e. the list does not change before and after this change.) Given that, I'd be happy to merge.

**Review by glessard — APPROVED** [review]
Thanks for putting in the effort to convert this to the newer syntax!

File: stdlib/public/core/Codable.swift [hunk 1]
```diff
@@ -5537,33 +5537,38 @@ internal struct _DictionaryCodingKey: CodingKey {
...
+
   @available(SwiftStdlib 5.6, *)
-  public var codingKey: CodingKey {
+  public var codingKey: any CodingKey {
     _DictionaryCodingKey(intValue: rawValue)
   }
+
   @available(SwiftStdlib 5.6, *)
-  public init?<T: CodingKey>(codingKey: T) {
+  public init?(codingKey: some CodingKey) {
```
Comment (lorentey) [inline stdlib/public/core/Codable.swift:5571]: Be careful with this one (and the two others of the same sort below) -- the constraints are weirdly split across the extension and the member, and rewriting them could easily end up causing a subtle ABI issue. This is exploring new language territory, and the ABI checker is unlikely to be of any help (other than flagging this as a potential break), but if the exported symbols match before/after this change, then we _may_ be okay.

**Review by lorentey — APPROVED** [review]
Lovely work! The danger of subtle ABI breaks is a bit higher than usual here, but if the exported symbol names match before/after this change, then we're probably fine.

(False positives in the ABI checker need to be reported to source analysis team, and added to the no-asserts allow list.)

File: stdlib/public/core/Codable.swift (hunk 1 above)
Comment (lorentey) [reply stdlib/public/core/Codable.swift:5571]: > I'm inclined to believe that api-digester has checked the mangled names. 

I'm not inclined to believe that. I don't think I've seen a list of symbols that libswiftCore is expected to export anywhere in this repository.

> However, the [macos.json](https://github.com/apple/swift/tree/main/utils/api_checker/FrameworkABIBaseline/Swift) baselines haven't been updated for 3 years, so they won't include the [CodingKeyRepresentable](https://github.com/apple/swift-evolution/blob/main/proposals/0320-codingkeyrepresentable.md) protocol.

D'oh. OK, so the ABI checker results are irrelevant, and we need to do manual checks.

Cc @nkcsgexi

**Pull request merged**